### ISSUE TITLE
[ENHANCEMENT] Infinite LR Schedules

### ISSUE THREAD

**Issue by Quentin-Anthony** [Zyphra/Megatron-LM#1]
We want support infinite LR schedules in Megatron-LM so that we can add more tokens during pre-training on-the-fly.

Some example schedules:

# Infinite LR Schedules
## Multiple Cosine warmup and decays 
![infLR_multiple_cosine_warmup](https://github.com/Zyphra/Megatron-LM/assets/10281105/bdc9291c-f396-4312-8bc5-5a743a31ceb4)

## Warmup + Constant LR + Exponential Decay
![infLR_constant_exp](https://github.com/Zyphra/Megatron-LM/assets/10281105/0e1c1419-9d91-42f3-a64a-bd7ae2f7f35b)

## Warmup + linear cool down + Constant LR + Exponential Decay
![infLR_lin_const_exp](https://github.com/Zyphra/Megatron-LM/assets/10281105/8708e011-8733-448a-8d68-10d1c4e3d7a2)

## Warmup + Inverse square root decay + Exponential decay
![infLR_invsqrt_exp](https://github.com/Zyphra/Megatron-LM/assets/10281105/3ec29ebf-c0d9-4115-a670-c800e2950a5b)

# Summary and key takeaways:
- Smoother is better (e.g. linear warmup + inv sqrt + exponential cooldown)
- https://arxiv.org/pdf/1910.10683.pdf
- https://arxiv.org/pdf/2106.04560.pdf
- Can also apply: https://arxiv.org/pdf/2308.04014.pdf 

I'm of the opinion that Warmup + Inverse square root decay + Exponential decay cooldown will be the best performing. The task is to complete the PR in https://github.com/Zyphra/Megatron-LM/pull/2 by porting its ipynb to Megatron's LR scheduler in https://github.com/Zyphra/Megatron-LM/blob/5f2877d85cb26e47ce6dcdae4b80adf376abf4e8/megatron/optimizer_param_scheduler.py#L9 

I believe you may be able to make this even simpler and just build atop the available `lr_decay_style` options such as https://github.com/Zyphra/Megatron-LM/blob/5f2877d85cb26e47ce6dcdae4b80adf376abf4e8/megatron/optimizer_param_scheduler.py#L95
